text/0021-state-pruning-and-fast-syncing.md

```diff
@@ -0,0 +1,240 @@
+1. If the request is a hit, try to download the object, deserialize it and validate the hash. Should any of these operations fail, discard and continue. Otherwise return.
+1. Ask the network (typically a node is known that is required to have it) for the object in the same manner.
+
+In the same manner, an S3-compatible API can be employed. Since many S3 providers allow making stored objects public, it would be highly recommendable to use a similar structure for the keys, as a result, every object uploaded to S3 automatically becomes read-only fetchable by other nodes that have access to the S3 bucket in real-time.
+
+Nodes configured to keep the external store up-to-date will, upon receiving a new object, attempt to upload it to the store, by first checking if it exists already and uploading it if it does not.
+
+With these measures in place almost all currently persisted state can be backed up to a service provider. It is up to operators or the Casper organisation to maintain these stores, which can be hosted on decentralized services like [storaj](https://storj.io/), providers like Amazon, Google, or self-hosted, possibly private caches like [minio](https://min.io/).
+
+### Transfering state from the trie store
+
+Objects like blocks, deploys and similar are trivially convertable into objects. For the trie store it should also be possible, as every vertex on the tree is either a leaf, a node or an extension. The hash matches what the store calls a pointer, thus a recovery of a root hash `4813494D137E1631BBA301D5ACAB6E7BB7AA74CE1185D456565EF51D737677B2` from an external store or the network would look like this:
+
+0. Initialize the retrieval queue with `4813494D137E1631BBA301D5ACAB6E7BB7AA74CE1185D456565EF51D737677B2`.
+0. For every object in the retrieval queue, remove those already in the trie store. If the queue is now empty, exit.
+0. Request all objects though the trie scheme, e.g. `triev1_4813494D137E1631BBA301D5ACAB6E7BB7AA74CE1185D456565EF51D737677B2`, which should deserialize to a `Trie::Node`.
+0. After successful deserialization, add all descendants of retrieved object into the retrieval queue, or abort with an error.
+0. Go to 1.
+
+### Optimization of retrieval
+
+Retrieval of any object can be optimized by batching, which will not be discussed in depth here. The general idea is to request multiple objects in a single request, and/or provide pre-assembled pack files that group data that is often requested at the same time, like a block and its deploys or the children of a subtree of a global state hash.
+
+This is an optimization that should be deferred for now though.
+
+### Putting everything together
+
+With a combination of the proposals above, a node's joining process from a trusted hash ideally becomes this:
+
+1. Download the block given in the trusted hash, sped-up through external caches.
```

Comment:
Is it necessarily faster?  Even if we're connected to a peer on a local network say?

Reply:
I'd say yes. The peer-on-local network is likely the exception for most operators. If so, this could be special cased or configurable.

Anyone else will find it hard to compete with a commercial CDN (which depending on your ISP might also be on the local network, although not financially :))